### DISCUSSION TITLE
How is dropdown work within html form ?

### DISCUSSION THREAD

**Discussion by sakhunzai** [mostafizurhimself/admintoolkit-html#177] [Q&A] [answered]
First of all thanks for great admin toolkit one of the best so far w.r.t refactoring and customization. I am in process of integrating this into my existing design and came across the issue of adding dropdown within html forms where we need to submit data to server. I dont see any example where the custom dropdown you provided is able to post data to server just like html dropdowns. 

Is it even possible to post data using [dropdown ](https://www.getadmintoolkit.com/dropdown.html). If its possible please add an example to demostrate it, it will be very kind , thanks.

**Comment by mostafizurhimself** [comment]
Please share some screenshots of your form to demonstrate your idea.

**Comment by sakhunzai** [comment]
I am taking the language dropdown from header section of template wrapped in a form element  and I want to submit the newly selected language to server whenever user selects new language , as given below.
>   
    
         <form action="http://www.foo.com" method="POST">
            <div class="dropdown" data-strategy="absolute">
              <div class="dropdown-toggle px-3">
                <button type="button"
                  class="inline-flex items-center gap-2 text-sm text-slate-600 hover:text-slate-700 focus:text-slate-700 dark:text-slate-400 dark:hover:text-slate-300 dark:focus:text-slate-300">
                  <span class="fi fi-gb"></span>
                  <span class="hidden font-medium md:inline-block">English</span>
                  <span class="inline-block font-medium md:hidden">EN</span>
                </button>
              </div>

              <div class="dropdown-content mt-3 w-40">
                <ul class="dropdown-list">
                  <li class="dropdown-list-item">
                    <button class="dropdown-btn" type="button">
                      <span class="fi fi-gb"></span>
                      <span class="">English</span>
                    </button>
                  </li>
                  <li class="dropdown-list-item">
                    <button class="dropdown-btn" type="button">
                      <span class="fi fi-de"></span>
                      <span class="">German</span>
                    </button>
                  </li>
                  <li class="dropdown-list-item">
                    <button class="dropdown-btn" type="button">
                      <span class="fi fi-gf"></span>
                      <span class="">French</span>
                    </button>
                  </li>
                  <li class="dropdown-list-item">
                    <button class="dropdown-btn" type="button">
                      <span class="fi fi-sa"></span>
                      <span class="">Arabic</span>
                    </button>
                  </li>
                  <li class="dropdown-list-item">
                    <button class="dropdown-btn" type="button">
                      <span class="fi fi-cn"></span>
                      <span class="">Chinese</span>
                    </button>
                  </li>
                  <li class="dropdown-list-item">
                    <button class="dropdown-btn" type="button">
                      <span class="fi fi-bd"></span>
                      <span class="">Bangla</span>
                    </button>
                  </li>
                </ul>
              </div>
            </div>
         </form>

**Comment by mostafizurhimself** [comment, marked as answer]
@sakhunzai In your case the suitable component is `select` component. Because `dropdown` components are not meant to be use as a form element. 
But if you want to use the `dropdown` instead, you have to use some little bit of javascript to handle the scenario. You can check this implementation  [theme-switcher.js](https://github.com/mostafizurhimself/admintoolkit-html/blob/main/src/js/components/theme-switcher.js) where we have used the dropdown to toggle theme.

**Comment by sakhunzai** [comment]
yes it looks like I need to tweak it , thanks. I ll update the thread about my solution

**Comment by sakhunzai** [comment]
So I have to do something like this:
> 
           
            <div class="dropdown" data-strategy="absolute" data-selected="EN" data-renderfn="renderLanguageItem" data-values='{"EN":["English","fi fi-gb"], "DE":["German","fi fi-de"]}'>
              <div class="dropdown-toggle px-3">
                <form  method="GET">
                  <input type="hidden" name="ln" value=""/>
                  <button type="button"
                    class="language-value inline-flex items-center gap-2 text-sm text-slate-600 hover:text-slate-700 focus:text-slate-700 dark:text-slate-400 dark:hover:text-slate-300 dark:focus:text-slate-300">
                    <span class="icon"></span>
                    <span class="name hidden font-medium md:inline-block"></span>
                    <span class="isocode inline-block font-medium md:hidden"></span>
                  </button>
                </form>
                <template class="item-template">
                  <li class="dropdown-list-item">
                    <button class="dropdown-btn" type="button">
                      <span class="icon "></span>
                      <span class="name"></span>
                      <span class="isocode"></span>
                    </button>
                  </li>
                </template>
                <script type="text/javascript">
                  function renderLanguageItem(dropdown){
                    if(dropdown.options.values){
                      let values=null;
                      try{
                        values=JSON.parse(dropdown.options.values);
                        dropdown.values=values;
                      }catch(e){
                        throw new Error('Unable to parse values');
                      }

                      let template = dropdown.target.querySelector('.item-template');
                      let items = dropdown.target.querySelector('.dropdown-list');
                      if(!template){
                        throw new Error('Unable to add items as template missing');
                      }else{
                        for(let isocode in values){
                          let item = template.content.cloneNode(true);
                          item.querySelector('.icon').classList.add(...values[isocode][1].split(' '));
                          item.querySelector('.name').textContent=values[isocode][0];
                          item.querySelector('.isocode').textContent=isocode;
                          item.querySelector('.isocode').classList.add('md:hidden');
                          item.querySelector('.dropdown-btn').value=isocode;
                          item.querySelector('.dropdown-btn').addEventListener('click',()=>{
                             selectItem(dropdown,isocode, values[isocode]);
                             dropdown.toggle.click();
                          });
                          items.appendChild(item);
                          if(isocode==dropdown.options.selected){
                            selectItem(dropdown, isocode, values[isocode])
                          }
                        }
                      }
                    }
                  }

                  function selectItem(dropdown,isocode, data){
                    let submit =false;
                    let inputBtn = dropdown.target.querySelector('.language-value')
                    if(inputBtn.value!=''){
                      let selectedItem=dropdown.values[dropdown.options.selected];
                      dropdown.target.querySelector('.icon').classList.remove(...selectedItem[1].split(' '));
                      submit =true;
                    }
                    dropdown.target.querySelector('.icon').classList.add(...data[1].split(' '));
                    dropdown.target.querySelector('.name').textContent=data[0];
                    dropdown.target.querySelector('.isocode').textContent=isocode;
                    inputBtn.value=isocode;
                    inputBtn.form.elements['ln'].value=isocode;
                    if(submit){
                      inputBtn.form.submit();
                    }
                  }
              </script>
              </div>

              <div class="dropdown-content mt-3 w-40">
                <ul class="dropdown-list">
                </ul>
              </div>
            </div>
            


with button component changes adding this at then end of init() function in  Dropdown component

```
//render template for given values
    if(this.options.renderfn){
      if(isFunction(window[this.options.renderfn]))
        window[this.options.renderfn](this)
      else
        throw new Error(`Dropdown item render function '${this.options.renderfn}' not found`);
    }
```
